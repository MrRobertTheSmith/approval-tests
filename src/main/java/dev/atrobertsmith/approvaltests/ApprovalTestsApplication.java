package dev.atrobertsmith.approvaltests;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class ApprovalTestsApplication {

	public static void main(String[] args) {
		SpringApplication.run(ApprovalTestsApplication.class, args);
	}

}
