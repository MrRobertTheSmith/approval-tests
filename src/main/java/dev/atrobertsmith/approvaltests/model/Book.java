package dev.atrobertsmith.approvaltests.model;

public record Book(String title, String author) { }
