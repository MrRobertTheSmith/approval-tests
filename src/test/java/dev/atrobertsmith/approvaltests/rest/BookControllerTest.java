package dev.atrobertsmith.approvaltests.rest;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.atrobertsmith.approvaltests.model.Book;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
class BookControllerTest {

    private ObjectMapper mapper;

    @Autowired
    private MockMvc mockMvc;

    @BeforeEach
    void setUp() {
        mapper = new ObjectMapper();
    }

    @Test
    void saveHappyPath() throws Exception {
        //GIVEN a book
        var book = mapper.writeValueAsString(new Book("Cat's Cradle", "K V"));

        //WHEN we save the book
        var saveResponse = mockMvc.perform(post("/api/v1/book")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(book))
                .andDo(print())
                .andExpect(status().is(200))
                .andReturn();

        //THEN
        Approvals.verify(saveResponse.getResponse().getContentAsString());
    }



    @Test
    void happyPathEndToEnd() throws Exception {
        //GIVEN a book is saved
        var book = mapper.writeValueAsString(new Book("Cat's Cradle", "K V"));

        mockMvc.perform(post("/api/v1/book")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(book))
                .andDo(print())
                .andExpect(status().is(200))
                .andReturn();

        //WHEN we fetch books
        var fetchResponse = mockMvc.perform(get("/api/v1/books"))
                .andDo(print())
                .andExpect(status().is(200))
                .andReturn();

        //THEN
        Approvals.verify(fetchResponse.getResponse().getContentAsString());
    }

}