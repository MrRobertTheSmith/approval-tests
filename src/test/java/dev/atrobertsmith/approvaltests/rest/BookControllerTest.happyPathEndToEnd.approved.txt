[{"title":"Cat's Cradle","author":"K V"},{"title":"Cat's Cradle","author":"K V"}]